### PR TITLE
Allow datatables to work within a repeater

### LINKED ISSUE #4063: Datatable Not Saved When Inside Repeater
- OctoberCMS Build: 446
- PHP Version: 7.1
- Database Engine: mysql
- Plugins Installed: 

### Description:
Adding a `datatable` form widget inside a `repeater` doesn't save the data.

### Steps To Reproduce:
https://github.com/octoberrain/test-plugin/pull/61

### PR DESCRIPTION
**Pull request by bennothommo**
Fixes #4063.

Allows datatable widgets (and other widgets) to enact their `getSaveValue` method inside a repeater, and fixes the POST data structure from a datatable to allow it to be repeated in a repeater.

### CODE REVIEW

**Comment by seanthepottingshed** [comment]
@bennothommo 

Excellent, the repeater datatable data does indeed get saved, however it would appear that the last repeater's datatable data get written over any previous repeater datatables

**Comment by bennothommo** [comment]
@seanthepottingshed Thanks for letting me know. I had that same issue - thought I had put some measures in place to stop it though.

Would you be able to send through a copy of the POST data that is sent in the AJAX call when saving your form with the repeating datatables, and perhaps a screenshot before save and after save so I can see what is occurring?

**Comment by seanthepottingshed** [comment]
@bennothommo 

Sure thing, see if the below helps:

**Before Save**
datatable-before-save

```
_session_key: Cj2yOLEv4oSQThzcMtY1DJXgaLgqjUYjxvVH0RHz
_token: 1UdDZI0QpcpfMOtWULgU1iJMJFuThYVY1QHUVgOM
Country[name]: Petoria
Country[code]: petoria
Country[is_active]: 0
Country[types]: 0
Country[language]: eh
Country[currency]: btc
Country[pages]: 0
Country[_page_error]: 
Country[states][0][title]: Stewie
Country[states][0][item][0][subtitle]: Suburb 1
___index_item[]: 0
Country[states][0][item][1][subtitle]: 
___index_item[]: 1
___index_states[]: 0
Country[states][1][title]: Brian
Country[states][1][item][0][subtitle]: 
___index_item[]: 0
Country[states][1][item][1][subtitle]: 
___index_item[]: 1
___index_states[]: 1
Country[states][2][title]: Chris
___index_states[]: 2
Country[states][3][title]: Lois
___index_states[]: 3
Country[states][4][title]: Meg
___index_states[]: 4
redirect: 0
safetyTableData[0][id]: 0
safetyTableData[0][label]: Suburb 2 Label 1
safetyTableData[0][rating]: 21
safetyTableData[1][id]: 1
safetyTableData[1][label]: Suburb 2 Label 2
safetyTableData[1][rating]: 22
```

**After Save**
datatable-after-save

**Comment by bennothommo** [comment]
@seanthepottingshed Would you mind checking again that you have tested it against my PR code? The POST data should look something similar to this - note that the table data starts with `Country[states][...][item][...][safety][TableData]` as opposed to `safetyDataTable[...]`.

```
_session_key: fcUpfo3AhutGHo7pj7DucqG45xr3dHklmLFAd3RF
_token: iI3QOiM15IH1ZZ7JsM78dlDW9I186JwvNAxP4RNL
Country[name]: Australia
Country[code]: australia
Country[is_active]: 0
Country[is_active]: 1
Country[types]: 0
Country[language]: 
Country[currency]: 
Country[pages]: 0
Country[_page_error]: 
Country[states][1][title]: WA
Country[states][1][item][1][subtitle]: Perth
___index_item[]: 1
Country[states][1][item][2][subtitle]: South Perth
___index_item[]: 2
___index_states[]: 1
Country[states][2][title]: SA
Country[states][2][item][3][subtitle]: Adelaide
___index_item[]: 3
___index_states[]: 2
Country[states][1][item][1][safety][TableData][0][id]: -1
Country[states][1][item][1][safety][TableData][0][label]: Perth Data 1
Country[states][1][item][1][safety][TableData][0][rating]: 11
Country[states][1][item][1][safety][TableData][1][id]: -2
Country[states][1][item][1][safety][TableData][1][label]: Perth Data 2
Country[states][1][item][1][safety][TableData][1][rating]: 22
Country[states][1][item][2][safety][TableData][0][id]: -1
Country[states][1][item][2][safety][TableData][0][label]: South Perth Data 1
Country[states][1][item][2][safety][TableData][0][rating]: 33
Country[states][1][item][2][safety][TableData][1][id]: -2
Country[states][1][item][2][safety][TableData][1][label]: South Perth Data 2
Country[states][1][item][2][safety][TableData][1][rating]: 44
Country[states][2][item][3][safety][TableData][0][id]: -1
Country[states][2][item][3][safety][TableData][0][label]: Adelaide Data 1
Country[states][2][item][3][safety][TableData][0][rating]: 55
```

If it still does not work, please try refreshing your browser cache and the October cache, just to be sure.

File: modules/backend/formwidgets/DataTable.php [hunk 1]
```diff
@@ -149,8 +149,13 @@ protected function makeTableWidget()
         $config = $this->makeConfig((array) $this->config);
 
         $config->dataSource = 'client';
-        $config->alias = studly_case(HtmlHelper::nameToId($this->fieldName)) . 'datatable';
-        $config->fieldName = $this->fieldName;
+        if (isset($this->parentForm->arrayName)) {
```
Comment (LukeTowers) [inline modules/backend/formwidgets/DataTable.php]: Could you use the getParentForm() method instead?

File: modules/backend/formwidgets/Repeater.php [hunk 2]
```diff
@@ -402,4 +418,83 @@ public function getGroupTitle($groupCode)
     {
         return array_get($this->groupDefinitions, $groupCode.'.name');
     }
+
+        /**
```
Comment (LukeTowers) [inline modules/backend/formwidgets/Repeater.php]: indentation

**Comment by LukeTowers** [comment]
@bennothommo could you add an example to the test plugin?

**Comment by bennothommo** [comment]
@LukeTowers: @seanthepottingshed has provided an example of this here: https://github.com/octoberrain/test-plugin/pull/61

**Comment by seanthepottingshed** [comment]
@bennothommo @LukeTowers 

Sorry for delay in getting back to  you, weekend and that. I will look into this now.

**Comment by seanthepottingshed** [comment]
@bennothommo @LukeTowers 

Sorry I'm not that experienced with testing pull requests, so I cloned october repo, added local DB, set `edgeUpdates` to true, and ran `git fetch origin pull/4102/head:repeating-datatable` - not sure if this how you're supposed to do it.

All working now, thank you so much it's really, really appreciated. When will this be merged into the next build?

**Comment by bennothommo** [comment]
@seanthepottingshed That's pretty much how I test PRs :)

**Comment by seanthepottingshed** [comment]
@bennothommo 

Excellent, just wanted to make sure I wasn't being one of the big thickos from the island!

[IMAGE ALT TEXT](https://youtu.be/fneTGPIU7xY?t=35 "The big thickos on the island")

**Pull request merged**